File: dsl-api/src/main/java/cbs/dsl/builder/TransactionBuilder.java
package cbs.dsl.builder;

import cbs.dsl.api.DslObject;
import cbs.dsl.api.ParameterDefinition;
import cbs.dsl.api.context.Context;
import cbs.dsl.api.context.TransactionContext;
import lombok.Getter;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.function.Consumer;
import java.util.function.Function;

/** Builder for creating transaction objects from DSL files. */
@Getter
public class TransactionBuilder {

  private final String code;
  private String name;
  private final List<ParameterDefinition> parameters = new ArrayList<>();
  private Function<Context, Context> contextBlock = Context::copy;
  private Function<TransactionContext, TransactionContext> previewBlock;
  private Function<TransactionContext, TransactionContext> executeBlock;
  private Function<TransactionContext, TransactionContext> rollbackBlock;

  TransactionBuilder(String code) {
    this.code = code;
  }

  public TransactionBuilder name(String name) {
    this.name = name;
    return this;
  }

  public TransactionBuilder parameters(Consumer<ParametersBuilder> block) {
    ParametersBuilder builder = new ParametersBuilder();
    block.accept(builder);
    this.parameters.addAll(builder.build());
    return this;
  }

  public TransactionBuilder context(Function<Context, Context> block) {
    this.contextBlock = block;
    return this;
  }

  public TransactionBuilder preview(Function<TransactionContext, TransactionContext> block) {
    this.previewBlock = block;
    return this;
  }

  public TransactionBuilder execute(Function<TransactionContext, TransactionContext> block) {
    this.executeBlock = block;
    return this;
  }

  public TransactionBuilder rollback(Function<TransactionContext, TransactionContext> block) {
    this.rollbackBlock = block;
    return this;
  }

  public List<ParameterDefinition> getParameters() {
    return Collections.unmodifiableList(new ArrayList<>(parameters));
  }

  public Function<Context, Context> context() {
    return contextBlock;
  }

  public Function<TransactionContext, TransactionContext> preview() {
    return previewBlock;
  }

  public Function<TransactionContext, TransactionContext> execute() {
    return executeBlock;
  }

  public Function<TransactionContext, TransactionContext> rollback() {
    return rollbackBlock;
  }

  public DslObject build() {
    return TransactionDslObject.builder()
        .code(code)
        .name(name)
        .parameters(Collections.unmodifiableList(new ArrayList<>(parameters)))
        .contextBlock(contextBlock)
        .previewBlock(previewBlock)
        .executeBlock(executeBlock)
        .rollbackBlock(rollbackBlock)
        .build();
  }
}

File: dsl/src/main/kotlin/cbs/dsl/runtime/TransactionBuilder.kt
package cbs.dsl.runtime

import cbs.dsl.api.ParameterDefinition
import cbs.dsl.api.TransactionDefinition
import cbs.dsl.api.context.TransactionContext
import cbs.dsl.api.context.TransactionDslContext
import cbs.dsl.api.context.TransactionPhase

class TransactionBuilder(override val code: String) : TransactionDefinition {
  private var _name: String? = null
  private val _parameters = mutableListOf<ParameterDefinition>()
  override val parameters: List<ParameterDefinition>
    get() = _parameters.toList()

  private var _contextBlock: (TransactionContext) -> Unit = {}
  override val contextBlock: (TransactionContext) -> Unit
    get() = _contextBlock

  private var _preview: ((TransactionContext) -> Unit)? = null
  private var _execute: ((TransactionContext) -> Unit)? = null
  private var _rollback: ((TransactionContext) -> Unit)? = null

  override val name: String?
    get() = _name

  val hasExecute: Boolean
    get() = _execute != null

  var delegateTarget: TransactionDefinition? = null

  override fun preview(ctx: TransactionContext) {
    _contextBlock(ctx)
    val dslCtx = TransactionDslContext(ctx, delegateTarget, TransactionPhase.PREVIEW)
    _preview?.invoke(dslCtx)
  }

  override fun execute(ctx: TransactionContext) {
    _contextBlock(ctx)
    val dslCtx = TransactionDslContext(ctx, delegateTarget, TransactionPhase.EXECUTE)
    _execute?.invoke(dslCtx) ?: error("Transaction '$code' has no execute block defined")
    ctx.enrichment.putAll(dslCtx.enrichment)
  }

  override fun rollback(ctx: TransactionContext) {
    _contextBlock(ctx)
    val dslCtx = TransactionDslContext(ctx, delegateTarget, TransactionPhase.ROLLBACK)
    _rollback?.invoke(dslCtx)
  }

  fun name(value: String) {
    _name = value
  }

  fun parameters(block: ParametersScope.() -> Unit) {
    _parameters += ParametersScope().apply(block).definitions
  }

  fun context(block: (TransactionContext) -> Unit) {
    _contextBlock = block
  }

  fun preview(block: (TransactionContext) -> Unit) {
    _preview = block
  }

  fun execute(block: (TransactionContext) -> Unit) {
    _execute = block
  }

  fun rollback(block: (TransactionContext) -> Unit) {
    _rollback = block
  }
}

fun transaction(code: String, block: TransactionBuilder.() -> Unit): TransactionDefinition =
    TransactionBuilder(code).apply(block)
